torch/csrc/jit/codegen/cuda/lower_utils.cpp

```diff
@@ -208,6 +227,60 @@ kir::Expr* applyReplacements(
   }
 }
 
+c10::optional<IterDomain*> getMaybeWarpReductionDim(
+    const kir::ReductionOp* node) {
+  auto kir_tv = ir_utils::getTVOutput(node);
+  if (!kir_tv) {
+    return c10::nullopt;
+  }
+  auto fuser_reduction = kir_tv->fuserTv()->definition()->as<ReductionOp>();
+  return getMaybeWarpReductionDim(fuser_reduction);
+}
+
+c10::optional<IterDomain*> getMaybeWarpReductionDim(const ReductionOp* node) {
+  auto fuser_tv_out = node->out()->as<TensorView>();
+  auto fuser_tv_in = node->in()->as<TensorView>();
+
+  // only support reducing to registers for now.
+  if (fuser_tv_in->getMemoryType() != MemoryType::Local ||
+      fuser_tv_out->getMemoryType() != MemoryType::Local) {
+    return c10::nullopt;
+  }
+
+  IterDomain* reduction_on_xdim = nullptr;
+  for (auto id : fuser_tv_out->domain()->domain()) {
+    // Currently warp reduction only allows
+    //  serial and block.x parallel reductions
+    if (id->isReduction() && id->isParallelized()) {
+      if (id->getParallelType() == ParallelType::TIDx) {
+        reduction_on_xdim = id;
+      } else {
+        return c10::nullopt;
+      }
+    }
+  }
+  if (!reduction_on_xdim) {
+    return c10::nullopt;
+  }
+
+  if (!reduction_on_xdim->start()->isZeroInt()) {
+    return c10::nullopt;
+  }
+
+  if (reduction_on_xdim->hasPaddingToMultipleOfWarp()) {
+    return c10::optional<IterDomain*>(reduction_on_xdim);
+  }
+
+  if (reduction_on_xdim->extent()->isConstScalar()) {
+    auto extent_value = reduction_on_xdim->extent()->getInt().value();
+    if (extent_value % C10_WARP_SIZE == 0) {
```

Comment:
This has the same concern as we discussed on the indexing. The fact that the extent is 32 or a multiple of it doesn't guarantee `blockDim.x` is also 32 or a multiple. It just means `blockDim.x` is greater or equal to the extent. I think that'd be a problem for the warp reduction function. 

I'm fine to leave this problem as is for now as long as it's tracked as an issue. Indeed, this ambiguity of `blockDim` can also be another problem (#1052), and so we need some mechanism to determine if the extent is exact or not. We could address this problem together as a separate PR.

Reply:
Yes. I have added a few checks to guard against this case but I do agree that a more systematic approach to the extents will be helpful, and sure we could work on this in a follow up.